File: pkg/feature/domain/strategy_evaluator.go
// Copyright 2024 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package domain

import (
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"strconv"

	"github.com/bucketeer-io/bucketeer/proto/feature"
)

const max = float64(0xffffffffffffffff)

type strategyEvaluator struct {
}

func (e *strategyEvaluator) Evaluate(
	strategy *feature.Strategy,
	userID string,
	variations []*feature.Variation,
	featureID string,
	samplingSeed string,
) (*feature.Variation, error) {
	switch strategy.Type {
	case feature.Strategy_FIXED:
		return findVariation(strategy.FixedStrategy.Variation, variations)
	case feature.Strategy_ROLLOUT:
		variationID, err := e.rollout(strategy.RolloutStrategy, userID, featureID, samplingSeed)
		if err != nil {
			return nil, err
		}
		return findVariation(variationID, variations)
	}
	return nil, errUnsupportedStrategy
}

func (e *strategyEvaluator) rollout(
	strategy *feature.RolloutStrategy,
	userID, featureID, samplingSeed string,
) (string, error) {
	bucket, err := e.bucket(userID, featureID, samplingSeed)
	if err != nil {
		return "", err
	}
	sum := 0.0
	for i := range strategy.Variations {
		sum += float64(strategy.Variations[i].Weight) / 100000.0
		if bucket < sum {
			return strategy.Variations[i].Variation, nil
		}
	}
	return "", errVariationNotFound
}

func (e *strategyEvaluator) bucket(userID string, featureID string, samplingSeed string) (float64, error) {
	hash := e.hash(userID, featureID, samplingSeed)
	// use first 16 characters (hex string) / first 8 bytes (byte array)
	intVal, err := strconv.ParseUint(hex.EncodeToString(hash[:])[:16], 16, 64)
	if err != nil {
		return 0.0, err
	}
	return float64(intVal) / max, nil
}

func (e *strategyEvaluator) hash(userID string, featureID string, samplingSeed string) [16]byte {
	// concat feature test id and user id
	// TODO: explain why this makes sense? Why does it make sense to add 'prerequisit' key here?
	concat := fmt.Sprintf("%s-%s%s", featureID, userID, samplingSeed)
	// returns 16 bytes which, if shown as hex string, has 32 characters
	return md5.Sum([]byte(concat))
}

File: evaluation-js/src/strategyEvaluator.ts
import * as crypto from 'crypto';
import { RolloutStrategy, Strategy } from './proto/feature/strategy_pb';
import { Variation } from './proto/feature/variation_pb';

const MAX = 0xffffffffffffffff;

class StrategyEvaluator {
  evaluate(
    strategy: Strategy,
    userID: string,
    variations: Variation[],
    featureID: string,
    samplingSeed: string
  ): Variation {
    switch (strategy.getType()) {
      case Strategy.Type.FIXED:
        return this.findVariation(strategy.getFixedStrategy()?.getVariation() || '', variations);
      case Strategy.Type.ROLLOUT:
        const variationID = this.rollout(strategy.getRolloutStrategy(), userID, featureID, samplingSeed);
        return this.findVariation(variationID, variations);
      default:
        throw new Error('Unsupported strategy');
    }
  }

  private rollout(
    strategy: RolloutStrategy,
    userID: string,
    featureID: string,
    samplingSeed: string
  ): string {
    const bucket = this.bucket(userID, featureID, samplingSeed);

    let sum = 0.0;
    for (const variation of strategy.getVariationsList()) {
      sum += variation.getWeight() / 100000.0;
      if (bucket < sum) {
        return variation.getVariation();
      }
    }
    throw new Error('Variation not found');
  }

  private bucket(userID: string, featureID: string, samplingSeed: string): number {
    const hash = this.hash(userID, featureID, samplingSeed);
    try {
      const intVal = parseInt(hash.slice(0, 16), 16);
      return intVal / MAX;
    } catch {
      throw new Error('Failed to calculate bucket value');
    }
  }

  private hash(userID: string, featureID: string, samplingSeed: string): string {
    const concat = `${featureID}-${userID}${samplingSeed}`;
    return crypto.createHash('md5').update(concat).digest('hex');
  }

  private findVariation(variationID: string, variations: Variation[]): Variation {
    for (const variation of variations) {
      if (variation.getId() === variationID) {
        return variation;
      }
    }
    throw new Error('Variation not found');
  }
}

export { StrategyEvaluator };